src/providers/Provider.py

```diff
@@ -34,22 +35,39 @@ def _get_dependencies(self, product: Product, *, get_dependencies: bool = True,
 
     def provide(self, products: List[Tuple[str, str]], concurrency: int = 1) -> List[Product]:
         cache = {}
+        cache_lock=Lock()
         all_products = []
-        current_products = [self._resolve_product(
-            name, version) for name, version in products]
-        with ThreadPool(concurrency) as pool:
-            while len(current_products) > 0:
-                # Update cache
-                for product in current_products:
-                    if(product.name not in cache):
-                        cache[product.name] = []
-                    if product.version in cache[product.name]:
-                        current_products.remove(product)
-                    else:
-                        cache[product.name] += [product.version]
-
-                all_products += current_products
-                deps = flatten(
-                    pool.map(self._get_dependencies, current_products))
-                current_products = pool.starmap(self._resolve_product, deps)
+        product_queue = Queue()
+
+        threads = [Thread(target=self.__provide_worker, args=[product_queue, all_products, cache, cache_lock], daemon=True)for _ in range(concurrency)]
+        for t in threads:
+            t.start()
+        
+        for product in products:
+            product_queue.put(product)
+        
+        # Wait for threads to finish and kill them
+        product_queue.join()
+        for _ in threads:
+            product_queue.put(None)
```

Comment:
It is unclear for me why you put `None`, is it because of line 59(in order to stop the iterator)? if so I would put it inside a function with an informative name

Reply:
It is. I'll put it in a function (or create a class for queue based thread pool?)

Reply:
A class with the proper doc is better imo

Reply:
Fixed

Reply:
@syncush ?